include/libtorrent/kademlia/dht_storage.hpp

```diff
@@ -0,0 +1,200 @@
+		// the data is returned inside the (entry) out parameter peers.
+		//
+		virtual bool lookup_peers(sha1_hash const& info_hash
+			, bool noseed, bool scrape
+			, entry& peers) const = 0;
+		// This method is named announce_peer for consistency with the
+		// upper layers, but has nothing to do with networking. Its only
+		// responsibility is store the peer in such a way that it's returned
+		// in the entry with the lookup_peers.
+		//
+		// The name parameter is the name of the torrent if provided during
+		// the announce_peer DHT message. The length of this value should
+		// have a maximum length in the final storage. The default
+		// implementation truncate the value for a maximum of 50 characters.
+		//
+		virtual void announce_peer(sha1_hash const& info_hash
+			, tcp::endpoint const& endp
+			, std::string const& name, bool seed) = 0;
+
+		// This function retrieves the immutable item given the target.
+		//
+		// For future implementers:
+		// The value should be returned as an entry in the key item["v"].
+		//
+		// returns true if the item is found and the data is returned
+		// inside the (entry) out parameter item.
+		//
+		virtual bool get_immutable_item(sha1_hash const& target
+			, entry& item) const = 0;
+		// Store the item's data. This layer is only for storage, the
+		// checking of the target info_hash and the data is performed in
+		// upper layers.
+		//
+		// For implementers:
+		// This data can be stored only if the target is not already
+		// present. The implementation should consider the value of
+		// dht_settings::max_dht_items.
+		//
+		virtual void put_immutable_item(sha1_hash const& target
+			, char const* buf, int size
+			, address const& addr) = 0;
+
+		// This function retrieves the sequence number of a mutable item.
+		//
+		// returns true if the item is found and the data is returned
+		// inside the out parameter seq.
+		//
+		virtual bool get_mutable_item_seq(sha1_hash const& target
+			, boost::int64_t& seq) const = 0;
+		// This function retrieves the mutable stored in the DHT.
+		//
+		// For implementers:
+		// The item sequence should be stored in the key item["seq"].
+		// if force_fill is true or (0 <= seq and seq < item["seq"])
+		// the following keys should be filled
+		// item["v"] - with the value no encoded.
+		// item["sig"] - with a string representation of the signature.
+		// item["k"] - with a string represnetation of the public key.
+		//
+		// returns true if the item is found and the data is returned
+		// inside the (entry) out parameter item.
+		//
+		virtual bool get_mutable_item(sha1_hash const& target
+			, boost::int64_t seq, bool force_fill
+			, entry& item) const = 0;
+		// Store the item's data. This layer is only for storage, the
+		// checking of the target info_hash and the valid signature is
+		// performed in upper layers.
+		//
+		// For implementers:
+		// The sequence number should be checked if the item is already
```

Comment:
it's unclear what "checked" means here. it looks a bit like a layer violation if a bug in the storage could result in a security issue. (i.e. failure to compare the sequence number)